core/deployment/src/main/java/io/quarkus/deployment/util/GenericTypeUtil.java

```diff
@@ -0,0 +1,160 @@
+package io.quarkus.deployment.util;
+
+import java.lang.reflect.Modifier;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+
+import org.jboss.jandex.ClassInfo;
+import org.jboss.jandex.ClassType;
+import org.jboss.jandex.DotName;
+import org.jboss.jandex.IndexView;
+import org.jboss.jandex.ParameterizedType;
+import org.jboss.jandex.Type;
+import org.jboss.jandex.TypeVariable;
+
+public final class GenericTypeUtil {
+
+    private static final DotName OBJECT = DotName.createSimple(Object.class.getName());
+
+    public GenericTypeUtil() {
+    }
+
+    /**
+     * Returns the captured generic types of an interface given a class that at some point in the class
+     * hierarchy implements the interface.
+     *
+     * The list contains the types in the same order as they are generic parameters defined on the interface
+     *
+     * A result is only returned iff all the generics where captured. If any of them where not defined by the class
+     * an exception is thrown
+     */
+    public static List<Type> getCapturedGenericTypesOfInterface(DotName clazz, DotName iface, IndexView index) {
+        final ClassInfo classInfo = index.getClassByName(clazz);
+        if (classInfo == null) {
+            throw new IllegalArgumentException("Class " + clazz + " was not found in the index");
+        }
+
+        final ClassInfo ifaceClassInfo = index.getClassByName(iface);
+        if (ifaceClassInfo == null) {
+            throw new IllegalArgumentException("Interface " + iface + " was not found in the index");
```

Comment:
I have a few other minor issues with the code but my main interrogation is: can we do it without requiring all the interfaces/superclasses to be in the index.

Because there's a good chance they will not be. Typically, `javax.validation.ConstraintValidator` for instance is not in the index. And I'm not sure I want to make it so.

AFAICS, without this preliminary check, you could live without the main interface being in the index or am I mistaken?

Reply:
Yes and do I would say.

Strictly speaking, yes the interface itself doesn't need to be in the index. But for the method to actually work, all the other interfaces in the path from the class to the interface do need to be in the index.

Why would `javax.validation.ConstraintValidator` not be in the index? In my case I look for Kubernetes resources which are not in the application index but they are in the combined index.

WDYT is the proper course of action here?

Reply:
Well, I'm wondering if we could support both classes and interfaces and remove this test entirely.

Even if we can't, I would remove this test.

All the classes are not in the index. I'm pretty sure the `javax.validation` ones won't be in the index except if I make them part of it.

Reply:
Supporting both classes and interfaces should be possible with a little more work. 
But I don't see a way of getting around things being in the index